### ISSUE TITLE
SynchronousFuture behavior is different depending on the SDK

### ISSUE THREAD

**Issue by johnpryan** [dart-lang/dart-pad#2030]
The behavior of a SynchronousFuture is different dependeing on whether the Dart SDK or Flutter SDK is used. The Dart SDK  executes immediately, but the Flutter SDK does not:

- [SynchronousFuture with Flutter SDK](https://dartpad.dev/?null_safety=true&id=16106aca0b944adbfc7a051a8933c247)
- [Synchronous Future with Dart SDK](https://dartpad.dev/?null_safety=true&id=92153b06d7fceddb8e9dd4c58a1809fd)

**Comment by Hixie** [comment]
Another reason they might differ is that in the Flutter case we wrap main() in a bunch of Zone magic.

**Comment by johnpryan** [comment]
It looks like this only happens when a Flutter app is compiled to JS:

| Platform      | sync | async |
|---------------|------|-------|
| Web (Dart)    | X    |       |
| Web (Flutter) |      | X     |
| VM (Dart)     | X    |       |
| VM (Flutter)  | X    |       |

FYI @yjbanov

**Comment by johnpryan** [comment]
Update: this is only happening with DDC.

**Comment by natebosch** [comment]
I'm surprised that we're able to subvert the asynchrony of `await` on any platform... 

@lrhn - were you aware of this behavior?

**Comment by lrhn** [comment]
It's unsurprising that things break when you implement `Future` incorrectly. Since `await` is defined in terms of calling `then`, and `then` is specified as *not* calling the callback during the `then` call, weird things happen if you do so anyway. We don't (necessarily) add extra overhead to try to avoid implementations of `Future` which are broken.
We also don't try to keep our implementations bug-compatible. If your `Future` doesn't behave like a proper future, we can go off the rails in different ways on different platforms.

Basically, it's unspecified behavior. If it hurts, stop doing it.

Nothing prevents you from creating a `Stream` which calls its event callbacks after you've called `cancel`, or after sending a done event, or sending more than one done event, or creating a future which calls the `then` callback more than once, or both the data and error callbacks. It breaks assumptions that other code might be based on, and the behavior of the platform is undefined if you do so. Same if your `Future` calls the callback in during the call to `then`.

**Comment by johnpryan** [comment]
Related to https://github.com/flutter/flutter/issues/82827

**Comment by parlough** [comment]
Thanks for the discussion here. Going to close in favor of the other linked issues as well as stale since all DartPad runs now consistently use DDC and the Flutter SDK.

If there's anything you'd like to see DartPad change still, please open a new issue with some more details. Thanks!
